javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -0,0 +1,455 @@
+        Objects.requireNonNull(that, "that is null");
+        Objects.requireNonNull(collisionResolution, "collisionResolution is null");
+        if (isEmpty()) {
+            return createFromEntries(that);
+        } else if (that.isEmpty()) {
+            return this;
+        } else {
+            Map<K, Traversable<V>> result = that.keySet().foldLeft(this.back, (map, key) -> {
+                final Traversable<V> thisValues = map.get(key).getOrElse((Traversable<V>) emptyContainer.get());
+                final Traversable<V2> thatValues = that.get(key).get();
+                final Traversable<V> newValues = collisionResolution.apply(thisValues, thatValues);
+                return map.put(key, newValues);
+            });
+            return createFromMap(result);
+        }
+    }
+
+    @Override
+    public Tuple2<Multimap<K, V>, Multimap<K, V>> partition(Predicate<? super Tuple2<K, V>> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final Tuple2<Iterator<Tuple2<K, V>>, Iterator<Tuple2<K, V>>> p = iterator().partition(predicate);
+        return Tuple.of(createFromEntries(p._1), createFromEntries(p._2));
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M peek(Consumer<? super Tuple2<K, V>> action) {
+        Objects.requireNonNull(action, "action is null");
+        if (!isEmpty()) {
+            action.accept(head());
+        }
+        return (M) this;
+    }
+
+    @Override
+    public String stringPrefix() {
+        return "Multimap[" + emptyMapSupplier().stringPrefix() + "," + emptyContainer.get().stringPrefix() + "]";
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M replace(Tuple2<K, V> currentElement, Tuple2<K, V> newElement) {
+        Objects.requireNonNull(currentElement, "currentElement is null");
+        Objects.requireNonNull(newElement, "newElement is null");
+        return containsKey(currentElement._1) ? (M) remove(currentElement._1).put(newElement) : (M) this;
+    }
+
+    @Override
+    public M replaceAll(Tuple2<K, V> currentElement, Tuple2<K, V> newElement) {
+        return replace(currentElement, newElement);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M retainAll(Iterable<? extends Tuple2<K, V>> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        return (M) createFromEntries(back.flatMap(t -> t._2.map(v -> Tuple.of(t._1, v))).retainAll(elements));
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public M scan(Tuple2<K, V> zero, BiFunction<? super Tuple2<K, V>, ? super Tuple2<K, V>, ? extends Tuple2<K, V>> operation) {
+        Objects.requireNonNull(operation, "operation is null");
+        return (M) Collections.scanLeft(this, zero, operation, Queue.<Tuple2<K, V>>empty(), Queue::append, this::createFromEntries);
```

Comment:
I had to look two times but I think using Queue is correct. The rear List is implicitly reversed when the finisher is called. Ok!